Documentation: Use Remark as linter in text editors (#159)
decaporg/decap-website: content/docs/editor-tooling.md

---
title: Editor Tooling
weight: 60

> group: Customization

Reviewer: @caseywatts, this is how you can redirect the old slug to this one in Hugo. [Hugo aliases docs](https://gohugo.io/methods/page/aliases/)

```suggestion
group: Customization
aliases:
  - custom-formatters
```

Author: great, thank you! 👏

---
title: Editor Tooling
weight: 60
group: Customization
aliases:
  - custom-formatters
---

## Linting via remark

If you want to lint your markdown files as part of your development workflow (in editors, in a pre-commit hook, and/or via CI), then we suggest you use `remark`. Remark is the library used by the Decap markdown widget.

> If you were to use a different linter than `remark` (such as `prettier`, etc), then you would end up with formatting conflicts between files. Files edited via the Decap web interface would be formatted differently than files edited via your development workflow.

Reviewer 2: Remark nor Prettier are linters by themselves, so I would reword it a bit for more clarity and to avoid a direct comparision.

What do you think about something like this?
```suggestion
If you want to lint your Markdown files as part of your development workflow through your code editor, pre-commit hooks, and/or CI, we recommend using the Remark ecosystem.

Since Decap CMS uses `remark` internally for Markdown processing, using it in your development workflow will help avoid formatting/style differences between files edited through Decap and files edited locally.
```

Author: I think this is clearer! 👏 

I think it would still be helpful to mention Prettier somewhere in the docs, as a search term (sorta like internal SEO) so people can find this page easily. I've seen several GitHub Issues that mention prettier on the Decap repo, so I think it would help people to find this

To set this up, you will need to install and set up both of these npm packages in your project:

- remark: https://remark.js.org/

> - remark-frontmatter: https://github.com/remarkjs/remark-frontmatter

Reviewer 2: I guess most of the time people will want to use remark-lint with remark-cli instead of the core `remark`? I know you can use it programatically through the `remark` API in a JS file, but we should focus on the easier path for people. What do you think about rewriting this part a bit so it leads users to using remark-lint instead? https://github.com/remarkjs/remark-lint#example-check-markdown-on-the-cli?

I think remark-lint will not mess with frontmatter, so maybe `remark-frontmatter` isn't even necessary? I haven't tested, if you could check this that would be great.

Author: I think you're right that people will likely use remark-cli instead of remark directly! That's what my team is doing, too. I can update that, no problem.

Author: On my project we did initially try to do it without `remark-frontmatter` and I wasn't able to get it working without that. The CLI interface scans each entire file, and it's not aware of the frontmatter enough to know how to skip it without that plugin.

In our Decap situation it behaves differently, because the markdown widget (which runs remark) is only manipulating the markdown body section of the markdown file. The Decap markdown widget is not aware of the frontmatter.

Author: One thing I noticed on the remark-lint page is that it suggests a bunch of plugins:

```
npm install remark-cli remark-preset-lint-consistent remark-preset-lint-recommended remark-lint-list-item-indent --save-dev
```

Since we want to make sure people are getting set up with the same configuration as the Decap markdown widget, maybe we could align these. Either:
1. Update the markdown widget to use this set of plugins (more in line with their current guidance, but it's a change we'd have to understand better & communicate out next version release)
2. Or suggest NOT using the plugins, and write a bit more configuration & notes here on this help page.

Which approach do you think would be better? (or do you have another idea?)

Reviewer 2: I'll say 2 should be easier, if you could bring a bit of what you did in your project as quite of a quick set up for remark linting would be great!

Author: Great -- I added the setup from my project as an example here 👏

---
title: Editor Tooling
weight: 60
group: Customization
aliases:
  - custom-formatters
---

> ## Linting via remark

Reviewer 2: I would probably just make the heading less specific so we can add other options in the future
```suggestion
## Linting
```

If you want to lint your Markdown files as part of your development workflow through your code editor, pre-commit hooks, and/or CI, we recommend using the Remark ecosystem.

Since Decap CMS uses `remark` internally for Markdown processing, using it in your development workflow will help avoid formatting/style differences between files edited through Decap and files edited locally.

To set this up, you will need to install and set up both of these npm packages in your project:

- remark-cli: https://github.com/remarkjs/remark/tree/main/packages/remark-cli
- remark-frontmatter: https://github.com/remarkjs/remark-frontmatter

Example installation command:

`npm install --save-dev remark-cli remark-frontmatter`

Example remark setup in `package.json`:

```json
{
  "scripts": {
    "remark:check": "remark .",
    "remark:write": "remark . --output"
  },
  "remarkConfig": {
    "plugins": ["remark-frontmatter"]
  }
}

> ```

Reviewer 2: Thanks @caseywatts, I did a pass at this section, let me know what you think!
````suggestion
To start using it in your project, install both [`remark-cli`](https://github.com/remarkjs/remark/tree/main/packages/remark-cli) and [`remark-frontmatter`](https://github.com/remarkjs/remark-frontmatter):

```sh
npm install --save-dev remark-cli remark-frontmatter
```

Then, add the following lines to your `package.json`:

```json
{
  "scripts": {
    "remark:check": "remark .",
    "remark:write": "remark . --output"
  },
  "remarkConfig": {
    "plugins": ["remark-frontmatter"]
  }
}
```

Now, you can use `npm run remark:check` to lint your Markdown files and `npm run remark:write` to format them.
````

Author: This looks great! I think it's really helpful to be opinionated like this, most people just want this sort of direction 🙏 

suggestion merged in ✅
